### PR TITLE
snowflake: Fix handling of `/~%` in the stage name

### PR DESCRIPTION
**Pull request by lustefaniak**
Fixes: #1008

Changed as discussed in the code review. Tokenizer accepts `/`, `%`, `~` normally, and then when parsing stage name, that is accepted as part of identifier.

### CODE REVIEW

File: tests/sqlparser_snowflake.rs [hunk 1]
```diff
@@ -1010,15 +1013,9 @@ fn test_snowflake_stage_object_names() {
...
-        ObjectName(vec![Ident::new("@namespace"), Ident::new("%table_name")]),
-        ObjectName(vec![
-            Ident::new("@namespace"),
-            Ident::new("%table_name/path"),
-        ]),
-        ObjectName(vec![
-            Ident::new("@namespace"),
-            Ident::new("stage_name/path"),
-        ]),
+        ObjectName(vec![Ident::new("@namespace.%table_name")]),
```
Comment (alamb) [inline tests/sqlparser_snowflake.rs:1016]: why were these tests changed?

**Review by alamb — COMMENTED** [review]
Thanks @lustefaniak  -- I have a question about why the tests were changed

Perhaps @pawel-big-lebowski can help review this as the author of  https://github.com/sqlparser-rs/sqlparser-rs/pull/841

**Comment by yuval-illumex** [comment]
Just wanted to say thanks to @lustefaniak - I also noticed this issue and wanted to tackle it.

File: tests/sqlparser_snowflake.rs (hunk 1 above)
Comment (lustefaniak) [reply tests/sqlparser_snowflake.rs:1016]: Tbh, I don't know how much big difference that makes, I don't even know if `.` in Snowflake makes a difference or not when checking the test cases. I was even thinking about keeping the test assertions as they were and just ignoring the tests until that gets fixed properly.

Comment (lustefaniak) [reply tests/sqlparser_snowflake.rs:1016]: I think that if `.` is a separator (assuming ObjectName was used here), then possibly we could keep `At`,`Mod` etc tokens, and somehow concat them to get SF identifiers, I don't know Snowlake that much to make a decision.

File: src/tokenizer.rs [hunk 2]
```diff
@@ -1004,6 +1004,18 @@ impl<'a> Tokenizer<'a> {
                             }
                         }
                         Some(' ') => Ok(Some(Token::AtSign)),
+                        // Snowflake stage identifier, this should be consumed as multiple dot separated word tokens
+                        Some(_) if dialect_of!(self is SnowflakeDialect) => {
+                            let mut s = "@".to_string();
+                            s.push_str(&peeking_take_while(chars, |ch| {
+                                self.dialect.is_identifier_part(ch)
+                                    || ch == '/'
```
Comment (pawel-big-lebowski) [inline src/tokenizer.rs:1012]: why is that removed from `is_identifier_part` method?

File: src/tokenizer.rs [hunk 3]
```diff
@@ -1004,6 +1004,18 @@ impl<'a> Tokenizer<'a> {
...
+                                    || ch == '~'
+                                    || ch == '%'
+                                    || ch == '.'
```
Comment (pawel-big-lebowski) [inline src/tokenizer.rs:1015]: I think this line is causing a problem and requires tests to be changed. 

The way I understand a problem is `/` can be either DIV token or part of identifier and this depends on a previous token (if a previous token before `.` started with `@`). I am not sure if this should be solved within tokenizer.  

Perhaps we should let it become DIV and then parse differently `indentifier div identifier` and check if the first identifier starts with `@`. Looks like a problem to be solved at the parsing layer.

Comment (yuval-illumex) [reply src/tokenizer.rs:1015]: @pawel-big-lebowski are there more places that '/' is part of identifier except of 'namespace/path'?
I'm not familiar with another places - maybe better approach will be to handle div in `parse_copy_into`?

Comment (pawel-big-lebowski) [reply src/tokenizer.rs:1015]: @yuval-illumex supporting `namespace/path` in `parse_copy_into` sounds good to me.

File: tests/sqlparser_snowflake.rs (hunk 1 above)
Comment (alamb) [reply tests/sqlparser_snowflake.rs:1016]: I am wary of changing tests when we don't understand the reason for why they changed

File: src/tokenizer.rs (hunk 3 above)
Comment (lustefaniak) [reply src/tokenizer.rs]: I redid the parsing so it tokenizes normally and then I collect all `@` `%` `/` as parts of Ident.

**Review by pawel-big-lebowski — APPROVED** [review]
Thanks for fixing this. 
Great job.

File: tests/sqlparser_snowflake.rs (hunk 1 above)
Comment (lustefaniak) [reply tests/sqlparser_snowflake.rs]: Updated PR to keep tokenising as before and instead add special handling in stage name parsing.

**Review by alamb — APPROVED** [review]
Thank you @lustefaniak for both this PR as well as https://github.com/sqlparser-rs/sqlparser-rs/pull/1025  -- they looks great.
